### DISCUSSION TITLE
Palette colors not maintained

### DISCUSSION THREAD

**Discussion by oli414** [ImageMagick/ImageMagick#5043] [Help] [answered]
I updated my ImageMagick from 7.0.8-6 Q16 to 7.1.0 Q16 today and noticed that the colors are different in the resulting output.

In the command, I run I have several remap actions to dither the image to different palettes. The results are masked so that parts of the image adhere to one palette, and other parts of the image adhere to the other.

This worked perfectly in the old version that I was using. And only the colors from my two palettes were used in the final image. 
The palette and main input images are all 8-bit. But the mask is created from a half-float (16-bit) EXR.

Unfortunately, after updating, some of the colors in the output file are off by 1 in the R, G, and/or B channels as a result of changing the color depth. This means that the output file now contains colors that were not on the palettes I remapped to.

![image](https://user-images.githubusercontent.com/2348094/163635022-8d46f219-6b3a-4eb2-9772-a2be0befd19f.png)

Does anyone happen to know what I might have to look into to get the command to work how it was with the older version again?

**Comment by fmw42** [comment]
What was your command line? What what platform/OS?

Did you specify the -dither?  There may have been a change in the default or the algorithm. If you did not specify the -dither, try adding that.

**Comment by oli414** [comment]
The platform is Windows 10, 64-bit.

This is what the dither section looks like:
`-dither FloydSteinberg -define dither:diffusion-amount=0% -remap "palette.png"`

The actual command is quite long (its generated by the application):
```
magick ( ( ( ( ( ".\output\.temp\meta_render_3.mpc" -write mpr:meta +delete ) ".\output\.temp\render_3.png" -write mpr:render )  -dither FloydSteinberg -define dither:diffusion-amount=0% -remap ".\palettes\default_vehicle_palette.png" ) ( mpr:render -dither FloydSteinberg -define dither:diffusion-amount=0% -remap ".\palettes\recolor_1_palette.png" ) ( mpr:meta -channel Green,Blue -evaluate set 0 +channel -fill "#00000000" +opaque "rgb(1,0,0)" -fill "#ffffffff" -opaque "rgb(1,0,0)" ) -composite ) ( mpr:render -dither FloydSteinberg -define dither:diffusion-amount=0% -remap ".\palettes\recolor_2_palette.png" ) ( mpr:meta -channel Green,Blue -evaluate set 0 +channel -fill "#00000000" +opaque "rgb(2,0,0)" -fill "#ffffffff" -opaque "rgb(2,0,0)" ) -composite ) ( mpr:meta ) -compose CopyOpacity -composite -bordercolor none -compose Copy -border 1 -trim  -format "%[fx:page.x - page.width/2] %[fx:page.y - page.height/2]" -write info: +repage ".\output\sprite_3.png"
```
The mask I described earlier is created from the 16-bit EXR image, the green and blue channels are discarded, and one specific red value is selected to mask out regions of the image that require a specific palette.

The input image does appear to be remapped properly, as the number of colors is reduced to the expected amount. It's just that some of them are a bit off.

**Reply by fmw42** [reply]
One of the IM developers will have to help. But I seem to recall some change to the dithering a while back.

**Reply by oli414** [reply]
Thanks for having a look, really appreciate it!

**Reply by fmw42** [reply]
I suggest that you do a simple test with only the remap operation to see if that is really the cause.

**Reply by oli414** [reply]
I did that, and that on its own did not show the issue. I suspect there's a rounding difference that occurs due to the combination of 16-bit and 8-bit files. But I'm not too sure

I'll continue to poke around with it in the mean time

**Comment by snibgo** [comment]
I suggest you insert "+write x0.png", "+write x1.png" etc before and after each "-remap" to isolate the problem. If, as you say, "the output file now contains colors that were not on the palettes I remapped to", that may provide a clue.

EDIT: You have:
```
-fill "#00000000" +opaque "rgb(1,0,0)"
```
This may make some pixels transparent black. Did you intend to make them transparent?

EDIT 2: You have "-composite" without specifying the "-compose" method. Your first file is an MPC, so it may have a "-compose" method attached. I suggest you change "-composite" to "-compose Over -composite".

EDIT 3: What is your platform? I assume Windows, but not Windows BAT.

**Reply by oli414** [reply]
> You have:

That is intentional

> You have "-composite" without specifying the "-compose" method. Your first file is an MPC, so it may have a "-compose" method attached. I suggest you change "-composite" to "-compose Over -composite".

Hm it seemed to have assumed Over, I'll add it just in case, it doesn't appear to affect much when I add it.

> What is your platform? I assume Windows, but not Windows BAT.

I run it through the Windows command line. The command is created in python and then ran using a subprocess.

**Comment by oli414** [comment]
After dissecting the command further I think I figured out where the problem is.

Testing the remap separately initially didn't show the issue, but it appears that I might have been using a palette file that was still generated with the old version of IM.

It looks like the palette I was using was not in fact 8-bit as I stated. But instead 64-bit.

This palette file is generated from another command that uses "montage". It seemed like the generated output was identical but upon closer inspection, this seems to have been the root of the issue, as the colors in the output file were different from the input files in this part of the process already.

The input files for the montage are mostly 24-bit PNGs, and one 32-bit (just a single transparent pixel) PNG.

`magick montage "input_a.png" "input_b.png" -tile x1 -geometry +0+0 -background none "palette.png"`

There is no reason for the input files for the montage to be any more than 8-bit. So I've reduced all of them to 8-bit. This creates a 32-bit image, so I force it to 8-bit on export. This seems to have solved my problem, and the commands now work consistently across versions.

I'm still not sure why the result was different in the new version. If I were to guess the rounding changed when going from a higher bit depth to a lower one.

**Reply by oli414** [reply]
Hm this resolved the issue for parts of the palette, but not all.

I still get incorrect colors from the "montage" command:

`magick montage a.png a.png -tile x1 -geometry +0+0 -background none PNG8:out.png`

This command just puts the same image next to each other.

Try it out on this file:

![a](https://user-images.githubusercontent.com/2348094/163684454-0cf8f6f8-8426-4b31-8023-ff3ee34b4fa5.png)

You'll see that the resulting image does not have the same colors:

![out](https://user-images.githubusercontent.com/2348094/163684470-4373e693-628a-407f-8e89-0c0ba5cb3a51.png)

**Comment by snibgo** [comment, marked as answer]
With IM v7.1.0-4, your montage command doesn't change any colours.

But you do get some small changes.

Why are you using "montage"? This is a complex command that does a "thumbnail" of each image, which resamples the images. If you simply want to append them:
```
magick a.png b.png c.png +append +repage out.png
```

**Reply by oli414** [reply]
Aha, that's it. I thought montage was what I needed, I didn't realize there was an easier way of doing it. Append works perfectly. Thanks for the explanation!
